### ISSUE TITLE
Calling (retry) in vcl_backend_error doesn't work if backend is responding with error

### ISSUE THREAD

**Issue by dugwood** [varnishcache/varnish-cache#4003]
### Expected Behavior

I've set the following:

```
sub vcl_backend_error
{
      std.syslog(3, "error: retries=" + bereq.retries + ", backend=" + bereq.backend);
      return (retry);
}
```

I've got a buggy `lighttpd` that fails on some specific POST requests. If it fails, Varnish don't execute retry. If I stop `lighttpd`, Varnish does the retries.

### Current Behavior

Varnish stops immediately, without doing any retries, from vcl_backend_error to vcl_synth.

### Steps to Reproduce (for bugs)

As this is a very specific issue (lighttpd crashes on POST when sent exactly at the keep-alive timeout), it's difficult to set a fully functional test. I don't even know what lighttpd is sending back.

### Context

Note that I use either fallback or round-robin directors, same issue. Even with a specific backend. So it's not linked to the backend selection.

I've added a `std.syslog()` in each `vcl_backend_*` function, and in `vcl_synth`.

When lighttpd is turned off (request is a POST too):
> systemd[1]: Stopped lighttpd.service - Lighttpd Daemon.
> systemd[1]: lighttpd.service: Consumed 1.352s CPU time.
> varnishd[36394]: fetch: retries=0, backend=local
> varnishd[36394]: error: retries=0, backend=local
> varnishd[36394]: fetch: retries=1, backend=local
> varnishd[36394]: error: retries=1, backend=local
> varnishd[36394]: fetch: retries=2, backend=local
> varnishd[36394]: error: retries=2, backend=local
> varnishd[36394]: fetch: retries=3, backend=local
> varnishd[36394]: error: retries=3, backend=local
> varnishd[36394]: fetch: retries=4, backend=local
> varnishd[36394]: error: retries=4, backend=local
> varnishd[36394]: synth!

When lighttpd fails on POST:
> varnishd[36394]: fetch: retries=0, backend=local
> varnishd[36394]: error: retries=0, backend=local
> varnishd[36394]: synth!

### Varnish Cache version

7.1.1

### Operating system

Debian 12

### Source of binary packages used (if any)

Debian official package

**Comment by dugwood** [comment]
Just tried latest version from Packagecloud:
> Debug: Version: varnish-7.4.1 revision d5a5aa9cc879320840ca467ddbb7df0f99c9ba0f

And same issue.

**Comment by Dridi** [comment]
The syslog output doesn't really help understand what is happening. The `varnishlog` output would give us a lot more context.

My best guess is that something fails in `vcl_backend_error` and prevents the retry from happening when lighttpd is present, so VCL execution is interrupted before reaching the `return (retry)` transition. It could be anything like a backend overflow, or possibly beresp headers overflow (not sure about this one, but maybe lighttpd is configured to add more headers and that tips over the `http_max_hdr` parameter).

Even with the `varnishlog` output we may not be able to understand what is happening without seeing some VCL, either of which can sometimes be difficult to share publicly.

You should probably inspect Varnish logs first.

**Comment by dugwood** [comment]
@Dridi I know it's kind of difficult to test...

`varnishlog` doesn't help, I forgot to attach it, but the log didn't help a bit:
```
*   << Request  >> 11        
-   Begin          req 4 rxreq
-   Timestamp      Start: 1697613014.856805 0.000000 0.000000
-   ReqProtocol    HTTP/2.0
-   Timestamp      Req: 1697613014.856807 0.000001 0.000001
-   VCL_use        boot
-   ReqStart       192.168.0.69 46336 a1
-   ReqMethod      POST
-   ReqURL         (removed)
-   ReqProtocol    HTTP/2.0
-   ReqHeader      host: (removed)
-   ReqHeader      scheme: https
-   ReqHeader      content-length: 124
-   ReqHeader      cache-control: max-age=0
-   ReqHeader      authorization: Basic eXZhbkBkdWd3b29kLmNvbTo=
-   ReqHeader      sec-ch-ua: "Chromium";v="118", "Google Chrome";v="118", "Not=A?Brand";v="99"
-   ReqHeader      sec-ch-ua-mobile: ?0
-   ReqHeader      sec-ch-ua-platform: "Linux"
-   ReqHeader      upgrade-insecure-requests: 1
-   ReqHeader      origin: null
-   ReqHeader      content-type: application/x-www-form-urlencoded
-   ReqHeader      user-agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36
-   ReqHeader      accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7
-   ReqHeader      sec-fetch-site: same-origin
-   ReqHeader      sec-fetch-mode: navigate
-   ReqHeader      sec-fetch-user: ?1
-   ReqHeader      sec-fetch-dest: document
-   ReqHeader      accept-encoding: gzip, deflate, br
-   ReqHeader      accept-language: fr-FR,fr;q=0.9,en;q=0.8,en-US;q=0.7,ca;q=0.6,de;q=0.5
-   ReqHeader      cookie: (removed)
-   ReqHeader      X-Forwarded-For: 192.168.0.69
-   ReqHeader      Via: 1.1 (removed) (Varnish/7.4)
-   VCL_call       RECV
-   ReqUnset       X-Forwarded-For: 192.168.0.69
-   ReqHeader      X-Forwarded-For: 192.168.0.69
-   ReqUnset       accept-encoding: gzip, deflate, br
-   ReqHeader      Accept-Encoding: gzip
-   ReqHeader      X-Environment: (removed)
-   ReqHeader      X-Forwarded-Host: (removed)
-   ReqUnset       host: (removed)
-   ReqHeader      host: (removed)
-   ReqUnset       X-Environment: (removed)
-   ReqHeader      X-Environment: (removed)
-   ReqHeader      X-Forwarded-Proto: https
-   ReqHeader      X-Engine: (removed)
-   ReqHeader      X-Pool: (removed)
-   VCL_return     pass
-   VCL_call       HASH
-   VCL_return     lookup
-   VCL_call       PASS
-   VCL_return     fetch
-   Link           bereq 32777 pass
-   Storage        malloc Transient
-   Timestamp      ReqBody: 1697613014.857437 0.000632 0.000630
-   Timestamp      Fetch: 1697613014.858052 0.001246 0.000614
-   RespProtocol   HTTP/1.1
-   RespStatus     503
-   RespReason     Service Unavailable
-   RespHeader     Date: Wed, 18 Oct 2023 07:10:14 GMT
-   RespHeader     Server: Varnish
-   RespHeader     X-Varnish: 11
-   VCL_call       SYNTH
-   RespHeader     Content-Type: text/html; charset=utf-8
-   RespHeader     Retry-After: 5
-   VCL_return     deliver
-   Timestamp      Process: 1697613014.858116 0.001310 0.000063
-   RespHeader     Content-Length: 708
-   Storage        malloc Transient
-   Filters        
-   RespProtocol   HTTP/2.0
-   Timestamp      Resp: 1697613014.858307 0.001501 0.000191
-   ReqAcct        3810 124 3934 98 708 806
-   End            
```

So my question would be: in what case `vcl_backend_error` won't do the `return (retry)`?

Of course I'll ask for a fix in lighttpd, but to me Varnish is buggy too. Any way to dump the lighttpd response within Varnish so that I can give you something to test with?

I can try to give you a minimum VCL, I'll clean mine from top to bottom to see if it's linked to something else.

**Comment by dugwood** [comment]
Here's a minimal VCL that still reproduce the issue, using Varnish 7.4.1:

```
vcl 4.1;
import std;
import directors;

/* Serveurs */
backend local {
        .host = "127.0.0.1";
        .port = "8080";
        .connect_timeout = 2s;
        .first_byte_timeout = 1800s;
        .between_bytes_timeout = 1800s;
}

sub vcl_recv {

                set req.http.X-Forwarded-Host = req.http.host;
                set req.http.host = regsub(req.http.host, "^dev\.", "");
                set req.http.X-Environment = "dev";

                set req.http.X-Forwarded-Proto = "https";

                set req.backend_hint = local;
                set req.http.X-Engine = "(removed)";
                set req.http.X-Pool = "(removed)";

        /* We only deal with GET and HEAD by default */
        if (req.method != "GET" && req.method != "HEAD")
        {
                return (pass);
        }

        return (hash);

}

sub vcl_backend_fetch {
        std.syslog(3, "fetch: retries=" + bereq.retries + ", backend=" + bereq.backend);

        return (fetch);
}

sub vcl_backend_response {
        std.syslog(3, "response: retries=" + bereq.retries + ", backend=" + bereq.backend);

        return (deliver);
}

sub vcl_deliver {
        std.syslog(3, "deliver!");

        return (deliver);
}
sub vcl_backend_error
{
        std.syslog(3, "error: retries=" + bereq.retries + ", backend=" + bereq.backend);

        return (retry);
}

sub vcl_synth {
        std.syslog(3, "synth!");
        set resp.body = {"<!DOCTYPE html>
        <head><title>"} + resp.status + {"</title>
        <body>
                Error "} + resp.status + {" - "} + resp.reason + {"
        </body>
        </html>"};

        return (deliver);
}

```

When hitting the lighttpd bug, I just see:
```
varnishd[9013]: fetch: retries=0, backend=local
varnishd[9013]: error: retries=0, backend=local
varnishd[9013]: synth!
```

**Comment by Dridi** [comment]
> Of course I'll ask for a fix in lighttpd, but to me Varnish is buggy too. Any way to dump the lighttpd response within Varnish so that I can give you something to test with?

I never claimed that the bug was in lighttpd, so far I suspect your setup to run into a failure mode, and I only came up with a potential cause explaining why lighttpd would trigger it.

You varnishlog output is not helpful, we don't see the backend transaction(s), you should try with `varnishlog -g request` to see the corresponding backend logs for your client request.

**Comment by dugwood** [comment]
@Dridi I'm sorry you've misinterpreted me, what I said it that lighttpd has a bug too. That's the origin of my discovery. If it didn't fail, then I would never have spotted this behaviour. As when lighttpd doesn't respond (network issue for instance), Varnish retries as expected. It's a combination of 2 specific issues.

I've extracted the varnishlog, tell me if it helps:
```
*   << Request  >> 65540     
-   Begin          req 65537 rxreq
-   Timestamp      Start: 1697615906.035470 0.000000 0.000000
-   ReqProtocol    HTTP/2.0
-   Timestamp      Req: 1697615906.035471 0.000001 0.000001
-   VCL_use        boot
-   ReqStart       192.168.0.69 52108 a1
-   ReqMethod      POST
-   ReqURL         (removed)
-   ReqProtocol    HTTP/2.0
-   ReqHeader      host: dev.(removed)
-   ReqHeader      scheme: https
-   ReqHeader      content-length: 124
-   ReqHeader      cache-control: max-age=0
-   ReqHeader      authorization: Basic (removed)
-   ReqHeader      sec-ch-ua: "Chromium";v="118", "Google Chrome";v="118", "Not=A?Brand";v="99"
-   ReqHeader      sec-ch-ua-mobile: ?0
-   ReqHeader      sec-ch-ua-platform: "Linux"
-   ReqHeader      upgrade-insecure-requests: 1
-   ReqHeader      origin: null
-   ReqHeader      content-type: application/x-www-form-urlencoded
-   ReqHeader      user-agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36
-   ReqHeader      accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7
-   ReqHeader      sec-fetch-site: same-origin
-   ReqHeader      sec-fetch-mode: navigate
-   ReqHeader      sec-fetch-user: ?1
-   ReqHeader      sec-fetch-dest: document
-   ReqHeader      accept-encoding: gzip, deflate, br
-   ReqHeader      accept-language: fr-FR,fr;q=0.9,en;q=0.8,en-US;q=0.7,ca;q=0.6,de;q=0.5
-   ReqHeader      cookie: (removed)
-   ReqHeader      X-Forwarded-For: 192.168.0.69
-   ReqHeader      Via: 1.1 (removed) (Varnish/7.4)
-   VCL_call       RECV
-   ReqHeader      X-Forwarded-Host: dev.(removed)
-   ReqUnset       host: dev.(removed)
-   ReqHeader      host: (removed)
-   ReqHeader      X-Environment: dev
-   ReqHeader      X-Forwarded-Proto: https
-   ReqHeader      X-Engine: core
-   ReqHeader      X-Pool: www
-   VCL_return     pass
-   VCL_call       HASH
-   VCL_return     lookup
-   VCL_call       PASS
-   VCL_return     fetch
-   Link           bereq 98307 pass
-   Storage        malloc Transient
-   Timestamp      ReqBody: 1697615906.035962 0.000492 0.000491
-   Timestamp      Fetch: 1697615906.036443 0.000973 0.000480
-   RespProtocol   HTTP/1.1
-   RespStatus     503
-   RespReason     Service Unavailable
-   RespHeader     Date: Wed, 18 Oct 2023 07:58:26 GMT
-   RespHeader     Server: Varnish
-   RespHeader     X-Varnish: 65540
-   VCL_call       SYNTH
-   VCL_return     deliver
-   Timestamp      Process: 1697615906.036495 0.001024 0.000051
-   RespHeader     Content-Length: 101
-   Storage        malloc Transient
-   Filters        
-   RespProtocol   HTTP/2.0
-   Timestamp      Resp: 1697615906.036633 0.001163 0.000138
-   ReqAcct        3810 124 3934 70 101 171
-   End            
**  << BeReq    >> 98307     
--  Begin          bereq 65540 pass
--  VCL_use        boot
--  Timestamp      Start: 1697615906.035807 0.000000 0.000000
--  BereqMethod    POST
--  BereqURL       (removed)
--  BereqProtocol  HTTP/2.0
--  BereqHeader    scheme: https
--  BereqHeader    content-length: 124
--  BereqHeader    cache-control: max-age=0
--  BereqHeader    authorization: Basic (removed)
--  BereqHeader    sec-ch-ua: "Chromium";v="118", "Google Chrome";v="118", "Not=A?Brand";v="99"
--  BereqHeader    sec-ch-ua-mobile: ?0
--  BereqHeader    sec-ch-ua-platform: "Linux"
--  BereqHeader    upgrade-insecure-requests: 1
--  BereqHeader    origin: null
--  BereqHeader    content-type: application/x-www-form-urlencoded
--  BereqHeader    user-agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36
--  BereqHeader    accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7
--  BereqHeader    sec-fetch-site: same-origin
--  BereqHeader    sec-fetch-mode: navigate
--  BereqHeader    sec-fetch-user: ?1
--  BereqHeader    sec-fetch-dest: document
--  BereqHeader    accept-encoding: gzip, deflate, br
--  BereqHeader    accept-language: fr-FR,fr;q=0.9,en;q=0.8,en-US;q=0.7,ca;q=0.6,de;q=0.5
--  BereqHeader    cookie: (removed)
--  BereqHeader    X-Forwarded-For: 192.168.0.69
--  BereqHeader    Via: 1.1 (removed) (Varnish/7.4)
--  BereqHeader    X-Forwarded-Host: dev.(removed)
--  BereqHeader    host: (removed)
--  BereqHeader    X-Environment: dev
--  BereqHeader    X-Forwarded-Proto: https
--  BereqHeader    X-Engine: core
--  BereqHeader    X-Pool: www
--  BereqProtocol  HTTP/1.1
--  BereqHeader    X-Varnish: 98307
--  VCL_call       BACKEND_FETCH
--  VCL_return     fetch
--  Timestamp      Fetch: 1697615906.035881 0.000073 0.000073
--  Timestamp      Connected: 1697615906.035886 0.000079 0.000005
--  BackendOpen    46 local 127.0.0.1 8080 127.0.0.1 57616 reuse
--  Timestamp      Bereq: 1697615906.035969 0.000162 0.000082
--  FetchError     HTC eof (-1)
--  BackendClose   46 local close Received bad req/resp
--  Timestamp      Beresp: 1697615906.036310 0.000502 0.000340
--  Timestamp      Error: 1697615906.036314 0.000506 0.000004
--  BerespProtocol HTTP/1.1
--  BerespStatus   503
--  BerespReason   Backend fetch failed
--  BerespHeader   Date: Wed, 18 Oct 2023 07:58:26 GMT
--  BerespHeader   Server: Varnish
--  VCL_call       BACKEND_ERROR
--  VCL_return     retry
--  Error          Retry not possible, req.body not cached
--  BereqAcct      5604 124 5728 0 0 0
--  End            
```

**Comment by walid-git** [comment]
Isn't this the cause of your issue ?
```
--  VCL_return     retry
--  Error          Retry not possible, req.body not cached
```

see: https://varnish-cache.org/docs/trunk/reference/vmod_std.html#std-cache-req-body

**Comment by dugwood** [comment]
Thanks @walid-git, so what I understand is that Varnish did get the error (as I go through `vcl_backend_error`), but as it's a POST request, and the POST's body isn't kept, it can't try the POST again on another backend?

So I need to save the POST's body so that I can replay the request?

**Comment by walid-git** [comment]
Right, exactly.

**Comment by dugwood** [comment]
So not a bug, a feature.

At least I've learnt about `varnishlog -g request`, that's useful for debugging a little deeper.

Closing it, thanks @Dridi and @walid-git :smile:

**Comment by Dridi** [comment]
I understand that it can be challenging to figure whether this is a bug in Varnish or in your setup, but when in doubt, please try the misc mailing list first.

https://varnish-cache.org/support/

**Comment by dugwood** [comment]
@Dridi of course I didn't read the Issue's disclaimer, but it states:

> If it's a feature request, please start a thread on the [varnish-misc](https://varnish-cache.org/support/index.html#mailing-lists) list instead.

So to me it was a bug, as I didn't had a way to « see » anything (before your hint on `-g request`).  Reading the code didn't help, neither the https://varnish-cache.org/docs/7.1/reference/states.html#reference-states which was a great finding too :-)

The fact that retries worked for GET and not for POST, and that the lighttpd bug is only for POST requests, didn't help focusing on « how Varnish handles a POST ». That mislead me :-)

**Comment by dugwood** [comment]
@Dridi I'm still trying to figure this out...

Based on what I've exchanged with lighttpd's main developer, it seems fine.

Say I have an HTTP/1.1 connection with keep-alive. I don't know when it will expire. So Varnish tries:

- first backend request: works, as there was no HTTP/1.1 channel
- second backend request: works, as the channel isn't closed (under the keep-alive timeout)
- third backend request, way past the keep-alive timeout: returns an error (`Connection: close` or something like this)

As it's a normal HTTP/1.1 behavior, the last call must be replayed, with a new connection to the backend.

As we've seen, I'm getting a `FetchError     HTC eof (-1)` which means that the channel was closed. So why Varnish doesn't retry it internally? Or does it?

In Debian 11, I was using Varnish 6.5.1 and lighttpd 1.4.59 (now I'm using respectively 7.1.1 and 1.4.69), and didn't see any issues on POSTs. Was there a major change between 6.5 and 7.1? I've only found https://github.com/varnishcache/varnish-cache/pull/3400 but that doesn't help much.

As I handle GET requests specifically with the `(retry)`, it never failed for GETs. But why does it fails for POSTs now...

Is there a way to get the reason of failure? So that if it's an HTTP/1.1 keep-alive timed out issue, I will replay the POST, and only in this case.

**Comment by asadsa92** [comment]
Could you give this patch a go and then share the logs again? https://github.com/varnishcache/varnish-cache/pull/4007

**Comment by asadsa92** [comment]
Also, a retry is not possible as mentioned unless you cache the reqbody using std.cache_req_body(), see https://varnish-cache.org/docs/trunk/reference/vmod_std.html#std-cache-req-body

**Comment by dugwood** [comment]
Thanks @asadsa92 I'll try the patch in a few hours.

I've understood why the POST isn't retried, but what I'm missing so far is why it worked before. Either lighttpd's HTTP/1.1 is different, or something has changed in Varnish.

Your patch will help, I'll let you know.

**Comment by dugwood** [comment]
@asadsa92 does this help?

```
**  << BeReq    >> 131075    
--  Begin          bereq 65544 pass
--  VCL_use        boot
--  Timestamp      Start: 1697646446.280222 0.000000 0.000000
--  BereqMethod    POST
--  BereqURL       (removed)
--  BereqProtocol  HTTP/2.0
--  BereqHeader    scheme: https
--  BereqHeader    content-length: 124
--  BereqHeader    cache-control: max-age=0
--  BereqHeader    authorization: Basic (removed)
--  BereqHeader    sec-ch-ua: "Chromium";v="118", "Google Chrome";v="118", "Not=A?Brand";v="99"
--  BereqHeader    sec-ch-ua-mobile: ?0
--  BereqHeader    sec-ch-ua-platform: "Linux"
--  BereqHeader    upgrade-insecure-requests: 1
--  BereqHeader    origin: null
--  BereqHeader    content-type: application/x-www-form-urlencoded
--  BereqHeader    user-agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36
--  BereqHeader    accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7
--  BereqHeader    sec-fetch-site: same-origin
--  BereqHeader    sec-fetch-mode: navigate
--  BereqHeader    sec-fetch-user: ?1
--  BereqHeader    sec-fetch-dest: document
--  BereqHeader    accept-language: fr-FR,fr;q=0.9,en;q=0.8,en-US;q=0.7,ca;q=0.6,de;q=0.5
--  BereqHeader    cookie: (removed)
--  BereqHeader    Via: 1.1 (removed) (Varnish/trunk)
--  BereqHeader    X-Forwarded-For: 192.168.0.69
--  BereqHeader    Accept-Encoding: gzip
--  BereqHeader    X-Forwarded-Host: dev.(removed)
--  BereqHeader    host: (removed)
--  BereqHeader    X-Environment: dev
--  BereqHeader    X-Forwarded-Proto: https
--  BereqHeader    X-Engine: core
--  BereqHeader    X-Pool: www
--  BereqProtocol  HTTP/1.1
--  BereqHeader    X-Varnish: 131075
--  VCL_call       BACKEND_FETCH
--  VCL_return     fetch
--  Timestamp      Fetch: 1697646446.280248 0.000025 0.000025
--  Timestamp      Connected: 1697646446.280277 0.000054 0.000029
--  BackendOpen    45 local 127.0.0.1 8080 127.0.0.1 52734 reuse 4.675430 6
--  Timestamp      Bereq: 1697646446.280478 0.000255 0.000200
--  FetchError     HTC eof (-1)
--  BackendClose   45 local close Received bad req/resp 4.676015 6
--  Timestamp      Beresp: 1697646446.281046 0.000823 0.000567
--  Timestamp      Error: 1697646446.281052 0.000829 0.000006
--  BerespProtocol HTTP/1.1
--  BerespStatus   503
--  BerespReason   Backend fetch failed
--  BerespHeader   Date: Wed, 18 Oct 2023 16:27:26 GMT
--  BerespHeader   Server: Varnish
--  VCL_call       BACKEND_ERROR
--  VCL_return     retry
--  Error          Retry not possible, req.body not cached
--  BereqAcct      5594 124 5718 0 0 0
--  End     
```

I've set the keep-alive timeout of lighttpd to 1 second so that I can trigger it easily.

**Comment by dugwood** [comment]
@Dridi @walid-git 

I've added « Connection: close » to the backend request (in `vcl_backend_fetch`), but connection is reused, which is wrong I think.

So far I'm going with:

- setting https://varnish-cache.org/docs/7.1/reference/varnishd.html#ref-param-backend-idle-timeout to 10 sec
- setting lighttpd Keep-Alive to 15 sec

As I currently have 60 sec for Varnish (default value), and 5 sec for lighttpd (default value), it's obvious lighttpd will end before Varnish. There's also a probability of lighttpd closing the connection after a number of requests (1000 as a default if I'm not mistaken).

**Comment by asadsa92** [comment]
Sorry, I  forgot to reply.

Yes, I think it make sense keep idle timeout of lighttpd > backend_idle_timeout, essentially you would never want the backend to close the connection on Varnish. Rather, you want Varnish to control the lifetime of the keep-alive connection.

We could, of course, add more VSL logs to make the troubleshooting easier.

**Comment by dugwood** [comment]
@asadsa92 @Dridi 

Also I've tried again:

> I've added « Connection: close » to the backend request (in vcl_backend_fetch), but connection is reused, which is wrong I think.

And it worked as expected (lighttpd doesn't create a keep-alive connection, so no bug in the POST).

I still have issues with `backend_idle_timeout` set to 60 and lighttpd set to 90s. I'll try with 30 and 90, that may help a bit.

And I've just found https://github.com/varnishcache/varnish-cache/issues/3633 which is clearly the same bug. But it seems to be linked to other timeouts (Varnish selecting a backend before receiving the client's full request).
